### PR TITLE
Add Applogy bidder

### PR DESCRIPTION
**Pull request by apaliy**
Add Applogy Bidder.

- Bidder interface implementation 

- Tests

- Configuration

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                        .body(body)
+                        .build()),
+                errors);
+    }
+
+    private MultiMap resolveHeaders() {
+        return HttpUtil.headers();
+    }
+
+    private ExtImpApplogy parseAndValidateImpExt(Imp imp) {
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: As I've previously noted - pass only `ObjectNode ext` from `imp.getExt()`. You don't need other `imp` fields here.

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        if (StringUtils.isBlank(extImpApplogy.getToken())) {
+            throw new PreBidException("token is empty");
+        }
+        return extImpApplogy;
+    }
+
+    private Imp processImp(Imp imp) {
+        Banner banner = imp.getBanner();
+        if (banner != null) {
+            if (banner.getH() == null && banner.getW() == null && CollectionUtils.isNotEmpty(banner.getFormat())) {
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: In go they also have a `== 0` check for banner Width and Height, and it should be a `||` (OR) condition, not AND.

Remove `CollectionUtils.isNotEmpty(banner.getFormat())` condition from here. We should check if they ARE empty before getting the first Format, and if they ARE empty - throw the exception - just like they have it in GO.

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid,
+                        requestImpIdToBidType.getOrDefault(bid.getImpid(), BidType.banner), DEFAULT_BID_CURRENCY))
+                .collect(Collectors.toList());
+    }
+
+    private static BidType getBidType(Imp imp) {
+        if (imp.getVideo() != null) {
+            return BidType.video;
+        } else if (imp.getXNative() != null) {
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: No need for `else if` when you have a return in previous condition. You can just keep 3 `if` statements, since each of them exits the method when `true`.

Also, you should throw/add an error when no bid type matched, just like they do in GO, and skip that bid. Check how I and @DGarbar did it in this PR - https://github.com/rubicon-project/prebid-server-java/pull/553/files in KubientBidder line 90 and 101, and check my second comment to his PR.

Basically, you need to pass an errors list and add an error to it whenever bid type was not matched, while returning a null instead of `BidderBid` object, so you can then filter these nulls out.

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || bidResponse.getSeatbid() == null) {
+            return Collections.emptyList();
+        }
+        return bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private static List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
+        final Map<String, BidType> requestImpIdToBidType = bidRequest.getImp().stream()
+                .collect(Collectors.toMap(Imp::getId, ApplogyBidder::getBidType));
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: This is not necessary, just have a method that iterates through `bidrequest.imp` and checks `imp.getId().equals(bid.getIpid())` and then applies the `getBidType` to resolve type.

Also, check the comments to `getBidType` below.

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid,
+                        requestImpIdToBidType.getOrDefault(bid.getImpid(), BidType.banner), DEFAULT_BID_CURRENCY))
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: This is incorrect, since they throw error when bid imp id was not matched to imp id, which makes using a map unnecessary. 

Check other `getBidType` or `resolveBidType` methods in other bidders, as well as check my comments to your `getBidType` method below.

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            return BidType.banner;
+
+        }
+    }
+
+    @Override
+    public Map<String, String> extractTargeting(ObjectNode ext) {
+        return Collections.emptyMap();
+    }
+}
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: It is quite annoying, but our project requires all `.java` files (except test files, but better have one there as well) to end up with an empty line.

Please be sure to have an empty line in all java files you add.

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/applogy/ExtImpApplogy.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+@AllArgsConstructor(staticName = "of")
+@Value
+public class ExtImpApplogy {
+
+    @JsonProperty("token")
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/applogy/ExtImpApplogy.java]: This annotation is redundant since the json field name can be resolved from variable name. Please remove it,

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/ExtImpRubicon.java [hunk 8]
```diff
@@ -32,4 +32,7 @@
     RubiconVideoParams video;
 
     String pchain;
+
+    @JsonProperty("is_rewarded_inventory")
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/ExtImpRubicon.java]: How the changes to Rubicon bidder are related to adding a new bidder? Please roll back this change and make a separate PR if this is not mistake and changes are required.

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/RubiconVideoParams.java [hunk 9]
```diff
@@ -21,4 +21,6 @@
     Integer skip;
 
     Integer skipdelay;
+
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/RubiconVideoParams.java]: Same as above - we are adding Applogy bidder, so why are we modifying Rubicon bidder?

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/VideoType.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package org.prebid.server.proto.openrtb.ext.request.rubicon;
+
+public enum VideoType {
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/VideoType.java]: Same as above

File: src/main/resources/bidder-config/applogy.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
+adapters:
+  applogy:
+    enabled: true
+    endpoint: http://localhost:8090
```
Comment (RossGoncharuk) [inline src/main/resources/bidder-config/applogy.yaml]: This is not where their AD Engine is) Check `config/config.go` from GO -> adapters.applogy.endpoint

As I understand - this and `enabled=true` are to run the Integration test. The config should contain actual bidder configuration and have bidder disabled by default, 

For test, add these properties to `src/test/resources/org/prebid/server/it/test-application.properties` which are used for `SpringBootTest` environment used in integration tests.

File: src/main/resources/bidder-config/applogy.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
+adapters:
+  applogy:
+    enabled: true
```
Comment (RossGoncharuk) [inline src/main/resources/bidder-config/applogy.yaml]: All new bidders are disabled by default.

File: src/test/resources/org/prebid/server/it/test-application.properties [hunk 13]
```diff
@@ -26,6 +26,10 @@ adapters.appnexus.enabled=true
 adapters.appnexus.endpoint=http://localhost:8090/appnexus-exchange
 adapters.appnexus.pbs-enforces-gdpr=true
 adapters.appnexus.usersync.url=//usersync-url/getuid?
+adapters.applogy.enabled=true
+adapters.applogy.endpoint=http://localhost:8090/applogy-exchange
+adapters.applogy.pbs-enforces-gdpr=true
+adapters.applogy.usersync.url=//usersync-url/getuid?
```
Comment (RossGoncharuk) [inline src/test/resources/org/prebid/server/it/test-application.properties]: Although it is not used, but please use a dummy-value like other bidders - `//applogy-usersync`. Let's keep up the project style.

**Comment by apaliy** [comment]
Completed requested changes

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+    @Before
+    public void setUp() {
+        applogyBidder = new ApplogyBidder(ENDPOINT_URL);
+    }
+
+    @Test
+    public void creationShouldFailOnInvalidEndpointUrl() {
+        assertThatIllegalArgumentException().isThrownBy(() -> new ApplogyBidder("invalid_url"));
+    }
+
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: The general approach (at least the one I use) is to either include all error cases first (as they appear in the method's flow) and then check all normal cases or go by methods flow combining error and regular cases. The first one is better and this is what we are using.

So here, please add a test case when there's an error while converting impression extension - check test named `makeHttpRequestsShouldReturnErrorIfImpExtCouldNotBeParsed` in other bidders' tests.

The key is to have all cases of possible inputs and resulting outputs covered to ensure all bidder aspects work as expected.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+                        .ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpApplogy.of(null)))));
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = applogyBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).hasSize(1)
+                .containsOnly(BidderError.badInput(
+                        "token is empty"));
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: No need to switch to next line, stay in one line where it is possible and doesn't harm code readability. Check other cases.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+                        "Applogy only supports banner, video or native ads"));
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .containsOnly(imp);
+    }
+
+
+    @Test
+    public void makeBidsShouldReturnErrorWhenResponseBodyCouldNotBeParsed() {
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: This should be placed after all `makeHttpRequests` method cases. Please ensure the correct tests order.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getBanner)
+                .extracting(Banner::getW, Banner::getH)
+                .containsOnly(tuple(300, 500));
+    }
+
+
+    @Test
+    public void makeBidsShouldReturnErrorIfResponseBodyCouldNotBeParsed() {
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: Duplicate of one you have above. Remove the one above as it is not placed where it should.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+
+        // then
+        assertThat(result.getErrors()).hasSize(1);
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("failed to decode json");
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_server_response);
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeBidsShouldReturnEmptyResultWhenResponseWithNoContent() {
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: Move test case down to where `makeBids` are located in accordance to the method's flow.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+        // then
+        assertThat(result.getErrors()).hasSize(1)
+                .containsOnly(BidderError.badInput(
+                        "Applogy only supports banner, video or native ads"));
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .containsOnly(imp);
+    }
+
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: Please add a case when banner doesn't have Width or Height and doesn't have the formats (imp.getFormat() == null) as well as when formats are empty ((imp.getFormat().size() == 0).

We need to check that "banner size information missing" exception is thrown as expected.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getBanner)
+                .extracting(Banner::getW, Banner::getH)
+                .containsOnly(tuple(300, 500));
+    }
+
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: Please add a test case to ensure the httpRequest URI being formed correctly by using the token from bidRequest's impression.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+        // when
+        final Result<List<BidderBid>> result = applogyBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .containsOnly(BidderBid.of(Bid.builder().impid("123").build(), video, "USD"));
+    }
+
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: Please add other cases like `makeBidsShouldReturnVideoBidIfVideoIsPresentInRequestImp` for cases when we have a Banner and Native to ensure that the bid type is resolved correctly.

File: src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+                mapper.writeValueAsString(BidResponse.builder().build()));
+
+        // when
+        final Result<List<BidderBid>> result = applogyBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).isEmpty();
+    }
+
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/bidder/applogy/ApplogyBidderTest.java]: Please add a test case for an error `"Failed to find impression %s", impId` when imp.id and bid.impid were not matched.

File: src/test/java/org/prebid/server/it/ApplogyTest.java [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+    @Test
+    public void openrtb2AuctionShouldRespondWithBidsFromApplogy() throws IOException, JSONException {
+        // given
+        // Applogy bid response for imp 001
+        wireMockRule.stubFor(post(urlPathEqualTo("/applogy-exchange/1234"))
+                .withHeader("Accept", equalTo("application/json"))
+                .withHeader("Content-Type", equalTo("application/json;charset=UTF-8"))
+                .withRequestBody(equalToJson(jsonFrom("openrtb2/applogy/test-applogy-bid-request-1.json")))
+                .willReturn(aResponse().withBody(jsonFrom("openrtb2/applogy/test-applogy-bid-response-1.json"))));
+        // Applogy bid response for imp 002
```
Comment (RossGoncharuk) [inline src/test/java/org/prebid/server/it/ApplogyTest.java]: Please add an empty line here to separate stubs for better code clarity and remove an extra empty line below, after expected response. Thanks!

**Comment by Hilson78** [comment]
Completed requested changes

**Comment by RossGoncharuk** [comment]
Also, please use a better naming for commit messages - they should be in imperative mood e.g. "Apply requested changes" or "Add Applogy bidder" and fit one line. Only exceptions are probably the merge commits made automatically by IDE. Check other PR's for examples.

File: src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java [hunk 24]
```diff
@@ -112,7 +90,7 @@ private Imp processImp(Imp imp) {
         Banner banner = imp.getBanner();
         if (banner != null) {
             if (banner.getH() == null || banner.getW() == null || banner.getH() == 0 || banner.getW() == 0) {
-                if (banner.getFormat() == null) {
+                if (banner.getFormat() == null || CollectionUtils.isEmpty(banner.getFormat())) {
```
Comment (RossGoncharuk) [inline src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java]: Please check the `CollectionUtils.isEmpty()` method. It is a null safe check of whether the given collection is empty, meaning that you don't need to check is passed collection is not null - that check is inside, otherwise you would just call `banner.getFormat().isEmpty()` after null check (which what that utility method does).

Just leave `CollectionUtils.isEmpty(banner.getFormat())` for the `if` condition.

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/ExtImpRubicon.java [hunk 25]
```diff
@@ -32,4 +32,5 @@
     RubiconVideoParams video;
 
     String pchain;
+
```
[the remark this answers is not in the document]
Comment (RossGoncharuk) [reply src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/ExtImpRubicon.java]: Have you rolled back Rubicon files?

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/applogy/ExtImpApplogy.java [hunk 26]
```diff
@@ -0,0 +1,12 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+@AllArgsConstructor(staticName = "of")
+@Value
+public class ExtImpApplogy {
+
+    String token;
+
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/applogy/ExtImpApplogy.java]: Please, remove empty line before class-closing curly brace.

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/ExtImpRubicon.java (hunk 25 above)
Comment (rpanchyk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/ExtImpRubicon.java]: Please, remove empty line before class-closing curly brace.

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/RubiconVideoParams.java [hunk 27]
```diff
@@ -21,4 +21,5 @@
...
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/rubicon/RubiconVideoParams.java]: Please, remove empty line before class-closing curly brace.
